pallets/deposits/src/lib.rs

```diff
@@ -0,0 +1,108 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+mod traits;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::pallet_prelude::*;
+	use frame_support::dispatch::fmt::Debug;
+
+	use frame_system::pallet_prelude::*;
+	use orml_traits::{MultiReservableCurrency, MultiCurrency, arithmetic::Bounded};
+	use common_types::{FundingType};
+	use crate::traits::{DepositCalculator, DepositHandler};
+	use codec::FullCodec;
+
+	type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+	type BalanceOf<T> = <<T as Config>::MultiCurrency as MultiCurrency<AccountIdOf<T>>>::Balance;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type MultiCurrency: MultiReservableCurrency<AccountIdOf<Self>, CurrencyId = Self::CurrencyId>;
+		/// The ID used to differentitate storage types.
+		type DepositId: FullCodec + Copy + Eq + PartialEq + Debug + MaxEncodedLen + TypeInfo;
+		/// The actual types that are being put in storage, abstracted as an enum;
+		type StorageItem: FullCodec + Eq + PartialEq + Copy + Debug;
+		/// The type responsible for calculating the cost of a storage item based on its type.
+		type DepositCalculator: DepositCalculator<BalanceOf<Self>, CurrencyId = Self::CurrencyId, StorageItem = Self::StorageItem>;
+		type CurrencyId: Clone + Copy + PartialOrd + Ord + PartialEq + Eq + Debug + Encode + Decode + TypeInfo + MaxEncodedLen;
+	}
+
+	/// A list of current deposits and the amount taken for the deposit.
+	#[pallet::storage]
+	pub type CurrentDeposits<T> = StorageMap<_, Blake2_128, <T as Config>::DepositId, Deposit<T>, OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
```

Comment:
We also need a method to slash deposits when a project fails its vote of no confidence, you can transfer the funds to both the treasury for now because there could be 10 contributors who are owed a portion of this deposit